### PR TITLE
Doc/git force

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

Twice this week I have written out an explanation of force-pushing so I wrote it up for the docs.

There is some minor re-arrangement of the sub-sections that I thought made more sense, but I would drop the first two commits with no argument if people prefer the old order.

### CODE REVIEW

File: doc/devel/development_workflow.rst [hunk 1]
```diff
@@ -358,3 +287,131 @@ and the history looks now like this::
...
+
+What is going on is that by default ``git push`` tries to protect you from
+accidentally make commits inaccessible by leaving no reference to them so if a
+push *would* effectively discard commits on the remote, ``git`` rejects the
+push.  When this happens GitHub adds the helpful suggestion to pull the remote
+changes and then try pushing again.  In some cases, such as if you and a
+colleague are both committing and pushing to the same branch, this is the
+correct course of action (or rebase your local branch but it is a matter of
+taste).  Although both of things are good ideas, taken together they are not
+what we want to do.
```
Comment (jklymak) [inline doc/devel/development_workflow.rst:401]:
```suggestion
What is going on is that by default ``git push`` tries to protect you from
accidentally discarding commits on the remote by rejecting the
push.  When this happens, GitHub adds the helpful suggestion to pull the remote
changes and then try pushing again.  In some cases, such as if you and a
colleague are both committing and pushing to the same branch, this is the
correct course of action (or rebase your local branch but it is a matter of
taste). 
```

File: doc/devel/development_workflow.rst [hunk 2]
```diff
@@ -358,3 +287,131 @@ and the history looks now like this::
...
+
+In the case of having intentionally re-written history we *want* to make the
+commits on GitHub inaccessible and replace them with the new-and-improved versions
+on our local machines.  In these cases what we want to do is ::
```
Comment (jklymak) [inline doc/devel/development_workflow.rst:405]:
```suggestion

However, in the case of having intentionally re-written history we *want* to discard the commits and  
replace them with the new-and-improved versions from our local branch.  In these cases what we want to do is ::
```

**Review by jklymak — COMMENTED** [review]
Some suggested tightening.  "Make commits inaccessible by leaving no reference..." is maybe conceptually different than "discarding commits" in some way I don't understand, but I don't think it's actually different?

File: doc/devel/development_workflow.rst (hunk 1 above)
Comment (tacaswell) [reply doc/devel/development_workflow.rst:401]: They are the same, but I was trying not to use the same slightly long phrase twice in the same sentence so I used two different overly long phrases in the same sentence.

Comment (jklymak) [reply doc/devel/development_workflow.rst]: I think for technical writing reusing the same phrase for a single concept is usually good.    I once wrote a paper where I called the same phenomena three different names - it did not go well in review ;-)

File: doc/devel/development_workflow.rst [hunk 3]
```diff
@@ -358,3 +287,128 @@ and the history looks now like this::
...
+What is going on is that by default ``git push`` tries to protect you from
+accidentally discarding commits on the remote by rejecting the
+push.  When this happens, GitHub adds the helpful suggestion to pull the remote
+changes and then try pushing again.  In some cases, such as if you and a
+colleague are both committing and pushing to the same branch, this is the
+correct course of action (or rebase your local branch but it is a matter of
+taste). 
+
+However, in the case of having intentionally re-written history we *want* to discard the commits and  
+replace them with the new-and-improved versions from our local branch.  In these cases what we want to do is ::
```
Comment (oscargus) [inline doc/devel/development_workflow.rst:402]:
```suggestion
replace them with the new-and-improved versions from our local branch.  In these cases, what we want to do is ::
```

File: doc/devel/development_workflow.rst [hunk 4]
```diff
@@ -358,3 +287,128 @@ and the history looks now like this::
...
+
+  $ git push --force-with-lease
+
+which tells git you are aware of the risks and want to do the push anyway.  We
+recommend using ``--force-with-lease`` over the ``--force`` flag.  The
+``--force`` will do the push no matter what, whereas ``--force-with-lease``
+will only do the push of the remote branch is where the local ``git`` client
```
Comment (oscargus) [inline doc/devel/development_workflow.rst:409]:
```suggestion
will only do the push if the remote branch is where the local ``git`` client
```

File: doc/devel/development_workflow.rst [hunk 5]
```diff
@@ -358,3 +287,128 @@ and the history looks now like this::
...
+to replace your already published commits with the new ones.
+
+
+.. _force-push:
+
+
+Pushing, with force
+-------------------
+
+If you have in some way re-written already pushed history (e.g. via
```
Comment (jklymak) [inline doc/devel/development_workflow.rst:379]: I wonder about adding a picture here as well, so it is clear what we are talking about.  I find the pictures above really useful.    You'd need a bit of description, but that seems straight forward, and we could refer to the remote branch as `origin` which I assume is what 99.9% of people call the remote.  

```
.. code-block:: none

                   A' cool-feature
                  /
     D---A---B---C origin/cool-feature
```

File: doc/devel/development_workflow.rst [hunk 6]
```diff
@@ -358,3 +287,128 @@ and the history looks now like this::
...
+while. In the meantime, ``main`` has progressed from commit E to commit (say) G:
+
+.. code-block:: none
+
+          A---B---C cool-feature
+         /
+    D---E---F---G main
+
+At this stage you consider merging ``main`` into your feature branch, and you
+remember that this here page sternly advises you not to do that, because the
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply doc/devel/development_workflow.rst:315]: I'll take the `here` change, but would like to defer other copy editing to a follow on PR.

File: doc/devel/development_workflow.rst [hunk 7]
```diff
@@ -358,3 +287,128 @@ and the history looks now like this::
...
+however if you *have* already pushed this commit see :ref:`force-push` for how
+to replace your already published commits with the new ones.
+
+
+.. _rebase-on-main:
+
+Rebasing on ``upstream/main``
+-----------------------------
+
+Let's say you thought of some work you'd like to do. You
```
[the remark this answers is not in the document]
Comment (oscargus) [reply doc/devel/development_workflow.rst:301]: (Not sure if I copy-edited the new or the old text...)

[the remark this answers is not in the document]
Comment (tacaswell) [reply doc/devel/development_workflow.rst:301]: You did some of both, but I took the trivial one to the old text.

File: doc/devel/development_workflow.rst (hunk 5 above)
Comment (tacaswell) [reply doc/devel/development_workflow.rst:380]: Adding the diagram but in the lines below so the diff did not make this auto resolve.

File: doc/devel/development_workflow.rst [hunk 8]
```diff
@@ -358,3 +287,141 @@ and the history looks now like this::
...
+versions from our local branch.  In this case, what we want to do is ::
+
+  $ git push --force-with-lease
+
+which tells git you are aware of the risks and want to do the push anyway.  We
+recommend using ``--force-with-lease`` over the ``--force`` flag.  The
+``--force`` will do the push no matter what, whereas ``--force-with-lease``
+will only do the push if the remote branch is where the local ``git`` client
+thought it was.
+
```
Comment (jklymak) [inline doc/devel/development_workflow.rst:424]:
```suggestion

After the force-push, the tree example will be the same on both the local and ``origin`` branches and the old commits, ``A, B, C``, will be lost (though hopefully squashed into ``A'``):

```
.. code-block:: none

      D---A'---E cool-feature, origin/cool-feature
```

Comment (jklymak) [reply doc/devel/development_workflow.rst:424]: Totally optional - I was about to merge, but thought it would be nice to the back to the example.

**Review by jklymak — APPROVED** [review]
I think we should merge.

It would be nice if there were an open-source definitive guide to git that we could just reference for all this stuff instead.   `gitwash` was an attempt at that, but the 5 or so git commands I _actually_ use were buried in pages of esoterica.  It seems a numfocus or scipy opinionated guide to the general git/github development flow that we could all just reference would be good.  We have one, xarray has a pretty good one, there are probably others.  But there is _nothing_ in here that is unique to Matplotlib.

**Review by melissawm — APPROVED** [review]
This looks good to me, thanks @tacaswell ! One thing that may be worth linking to, but could be on a follow up, is some documentation on reflog. I've found not a lot of people know about it and it can be really helpful when trying to understand what happened or in what state the repo is.

**Pull request merged**